src/sage/modular/drinfeld_modform/ring.py

```diff
@@ -0,0 +1,653 @@
+        TypeError: names must be None or a comma seperated string
+        sage: DrinfeldModularForms(Frac(GF(2)['T']), rank=3, names='f1, f2, f3, f4')
+        Traceback (most recent call last):
+        ...
+        ValueError: the number of generators must be equal to the rank (=3)
+    """
+
+    Element = DrinfeldModularFormsElement
+
+    @staticmethod
+    def __classcall_private__(cls, base_ring, rank=2, group=None,
+                              has_type=False, names=None):
+        rank = ZZ(rank)  # check the type of rank
+        if not isinstance(base_ring, FractionField_generic):
+            raise TypeError("base ring must be a fraction field of a "
+                            "polynomial ring")
+        if not isinstance(base_ring.base(), PolynomialRing_general):  # not sure if this test is relevant
+            raise TypeError("the base of the base ring must be a "
+                            "polynomial ring")
+        if not base_ring.characteristic():
+            raise ValueError("base ring characteristic must be finite")
+        if not base_ring.base().base().is_field():
+            raise ValueError("the ring of constants must be a field")
+        if not base_ring.base().base().is_finite():
+            raise ValueError("the ring of constants must be finite")
+        if group is not None:  # placeholder
+            raise NotImplementedError("Drinfeld modular forms are currently "
+                                      "only implemented for the full group")
+        if names is None: # default names
+            names = f"g1, "
+            for i in range(2, rank, 1):
+                names += f"g{i}, "
+            if has_type:
+                names += f"h{rank}"
+            else:
+                names += f"g{rank}"
+        elif isinstance(names, str):
+            nb_names = len(names.split())
+            if nb_names != rank:
+                raise ValueError("the number of generators "
+                                 f"must be equal to the rank (={rank})")
+        else:
+            raise TypeError("names must be None or a comma seperated string")
+        return cls.__classcall__(cls, base_ring, rank, group, has_type, names)
+
+    def __init__(self, base_ring, rank, group, has_type, names):
+        self._has_type = has_type
+        self._rank = rank
+        self._base_ring = base_ring
+        q = base_ring.base_ring().cardinality()
+        if has_type:
+            degs = [q**i - 1 for i in range(1, rank, 1)]
+            degs.append((q**rank - 1)/(q - 1))
+        else:
+            degs = [q**i - 1 for i in range(1, rank + 1, 1)]
+        self._poly_ring = PolynomialRing(base_ring, rank, names=names,
+                                         order=TermOrder('wdeglex', degs))
+        self._assign_names(names)
+
+        Parent.__init__(self, base=base_ring, category=GradedAlgebras(base_ring))
+
+    def _an_element_(self):
+        r"""
+        Return an element of self.
+
+        TESTS::
```

Comment:
Why put these in a `TESTS` and not `EXAMPLES`?
Btw, I noticed that `M.an_element?` returns the docstring of `Parent.an_element`:

```
sage: M.an_element?
Docstring:     
   Returns a (preferably typical) element of this parent.

   This is used both for illustration and testing purposes. If the set
   "self" is empty, "an_element()" raises the exception
   "EmptySetError".

   This calls "_an_element_()" (which see), and caches the result.
   Parent are thus encouraged to override "_an_element_()".

   EXAMPLES:

      sage: CDF.an_element()
      1.0*I
      sage: ZZ[['t']].an_element()
      t

   In case the set is empty, an "EmptySetError" is raised:

      sage: Set([]).an_element()
      Traceback (most recent call last):
      ...
      EmptySetError
Init docstring: Initialize self.  See help(type(self)) for accurate signature.
File:           ~/pro/code/sage/src/sage/structure/parent.pyx
Type:           builtin_function_or_method
sage: 
```

Reply:
The block here doesn't really matter because this method is hidden to the user and therefore it doesn't show in the reference manual.

Also, about the docstring, that's normal because `M.an_element` is implemented in `Parent`. To see the docstring of the Drinfeld modular form implementation, you need to type `M._an_element_?`